### ISSUE TITLE
Parameters for transformation scripts

### ISSUE THREAD

**Issue by hiker** [stfc/PSyclone#2757]
During optimisation (including profiling runs), I ended up using three or four different versions of a transformation script (`original and with tiling`  and `profile and non_profile`). This makes it sometimes hard to keep the scripts in synch.

It would be nice if options could be specified for the transformation scripts. This would allow me to have just one version of the script, and based on the command line add profiling or not (and maybe even add tiling or not).

For now I can use environment variables, but it might be a useful feature to be able to see these options in e.g. build log files.

An alternative approach might be to make transformation scripts call each other (e.g. the profiling script first calls the base script, then applies profiling to the output). But that still makes it hard to then have one profiling-function that will be added to the original and tiled version of the scripts ... unless we have user input.

**Comment by sergisiso** [comment]
I find using envvars a good solution (as you suggested), see: https://github.com/stfc/PSyclone/blob/33ff7e8991d0ce0d8c6e7d34ac1480684bc20da5/examples/lfric/scripts/gpu_offloading.py#L61

Also because sometimes the build system needs to adapt: link or not profiling flags, use offloading of threading flags, ... and it can read the same envvar.
Do you think we need a different mechanism?

**Comment by hiker** [comment]
There is one big advantage if the options are 'visible' on the PSyclone command line: Fab computes a hash code based on command line options (etc etc). So, changing an env variable will not change the hash code, and as a user you will have to be aware of this and trigger a clean recompile  (admittedly, you might as well have two different build trees?? But if you think of lfric, you might not want to have two completely different build tree just because one file changes).

Very easy to understand would be something like:
```
psyclone ... -s "my_script(PROFILE=1, TILING=3)"  or so.
```
We could then pass the arguments as kwargs (or options dictionary? That might make it easier to reuse scripts) to the user-script when calling it (in addition to the tree information). But agreed, no rush.

**Comment by sergisiso** [comment]
Note that the space (after the comma) will be problematic for an argument, which can be solved with quotes; and that the other symbols are possible in filenames (you can have a file named just like that), so this notation is ambiguous. But we can explore options, what about e.g. --kwargs "tile_size:4, tiling:yes"

**Comment by hiker** [comment]
Yes, I indeed forgot the quotes  ... hang on, no I didn't (at least not in this ticket, in the teams chat I did :)  ). I have to say that I prefer my version, simply because it makes it very obvious for a reader what this does, while `--kwargs` does not (maybe calling it `--script-kwargs`)? 

I am not sure I understand the problem with the ambiguity - yes, in theory a file could contain parenthesis, but this would be a case for me to prefer clear understanding of the meaning of the syntax vs a super-extremely-very-unlikely thing to ever happen.

**Comment by sergisiso** [comment]
> I am not sure I understand the problem with the ambiguity - yes, in theory a file could contain parenthesis, but this would be a case for me to prefer clear understanding of the meaning of the syntax vs a super-extremely-very-unlikely thing to ever happen.

The thing is that filenames and dict-syntax are well known and defined concepts in unix/python. So, we can give the filename to the open method and eval the dict string without having to create any parsing and if they are invalid proper error messages are generated. If we go for an own format, all parsing, validation, error messages, documentation, ... is up to us.

**Comment by hiker** [comment]
I have started to look at this issue, since it will work nicely together with the PSyclone support in Fab that I am doing (my goal being that the yaml files used in lfric can specify file-specific parameters).

I prefer to avoid `eval` (because of security/potential vulnerabilities - which might be an issue for our site, i.e. security might scan and flag scripts that use `eval`). I can parse the expressions using Python's ast module, and if we restrict parameters to literals (i.e. no expressions like `1+2`), we can convert them safely  (including lists , dicts etc. e.g.: `{'a': 1, 'b': {'c': 'd', 'e': 1}}` ).

I intend to only support keyword args (not positional ones) - I feel we agree on that :)

Having done this, both approaches (`global.py(...)` and `--script-kwargs ...`) are easy enough to implement (now that I have the parsing sorted out). I have started with 'my' way (`-s 'global.py(a="1")'`), but just realised one thing: it doesn't allow us to specify parameters for transformation of the algorithm layer. But command line parameters easily do. So, I might abandon my way (and cry a little bit :) ) and only support specification of script arguments using additional command line option(s). That still leaves a few approaches. We could use:

- `--psy-kwargs`
    For modifications of the psy-layer. 
- `--alg-kwargs`
    For modifications of the alg layer
- `--script-kwargs`
    for transmutation.

Or should we have just `--script-kwargs` (which would be used for psy-layer and transmutation), and `--alg-kwargs` (for algorithm transformation)?  Or we could just have one command line parameter (`--script-kwargs`), and the arguments would be provided to all scripts (alg/psy/transmute)? Or of course (just for completeness sake), we don't support alg-layer args at all, and leave it to just one command line option.

@sergisiso , @arporter , @LonelyCat124 - feedback welcome :)

**Comment by sergisiso** [comment]
I would go for a single option `--script-kwargs`, provided to each trans entry point.

I agree with the avoiding eval comment, and also in using something like `ast.literal_eval` as we don't need to evaluate expression, just input dictionary/list/tuple values.

**Comment by arporter** [comment]
#2757 is merged.
